rfcs/0017-intensional-store.md

```diff
@@ -0,0 +1,110 @@
+---
+feature: intensional_store
+start-date: 2017-08-11
+author: Wout.Mertens@gmail.com
+co-authors: (find a buddy later to help our with the RFC)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+One paragraph explanation of the feature.
+
+# Motivation
+[motivation]: #motivation
+
+* Better re-use of inputs between compiles
+* Faster updates via nix-channel
+* Less compiling
+* More benefit from reproducible compiles, so more reason to work on that
+
+# Detailed design
+[design]: #detailed-design
+
+Terms used:
+* derivation: a `nix-build` output product depending on some inputs and resulting in a file or directory under `/nix/store`
+* dependent derivation: a derivation built using the currently considered derivation
+* `$out`: name of the location where a derivation is built, e.g., `zyb6qaasr5yhh2r4484x01cy87xzddn7-unit-script-1.12`
+  * calculated based on the hashes of all the inputs, including build tools
+* `$cas`: output hash, the total hash of all the files under $out, with the derivation name appended, e.g., `qqzyb6bsr5yhh2r5624x01cy87xzn7aa-unit-script-1.12`
+
+## Concept
+
+The basic concept is aliasing equivalent input derivations in such a way that dependent derivations won't need to change if only `$out` changes but not the input derivation contents.
+
+After building a derivation, `$cas` is calculated, and `$out` is renamed to `$cas`. Then, if another build requires the input `$out`, it gets `$cas` instead, and all references to that build input will be `$cas` instead of `$out`. That dependent derivation will also have its input hash calculated with the `$cas` instead of the `$out`.
+
+This means that if 2 different derivations of the same input have a different `$out` but the same `$cas`, any dependent builds will not need to rebuild due to the inputs being different. For example, the 12MB input `poppler-data` is often the same across multiple different input derivations, so many `$out`s for `poppler-data` all result in the same `$cas`. Similarly, a compiler flag change might leave most derivations unchanged.
+
+In order to know which `$out`s refer to a particular `$cas`, symlinks can be used (`$out` pointing to `$cas`), or that data can be stored in the store database. The database can help with doing reverse lookups from `$cas` to all the `$out`s.
+
+## Calculating `$cas`
+
+There is one important corner case that needs special handling: if a derivation refers to itself, it will be referring to `$out`, because `$cas` is not known at the time of the build. This means that each `$out` of a furthermore equivalent build would have a different hash, due to the different `$out`s.
+
+To fix this, the `$cas` calculation has to replace all occurrences of `$out` with an equal-length string of (for example) NULL bytes. After that, `$out` is renamed to `$cas` and all occurrences of `$out` are replaced with `$cas`.
```

Comment:
@edolstra thesis already describe a way to handle that, which is already implement in Nix. i-e removing the hashes, and prepending the location of the hashes to hashed content.  The pre-pend it to avoid ambiguities, as far as I understand.

Reply:
Yes, this should just follow the thesis unless there's reason not to